src/Decompiler/Typing/ExpressionTypeAscender.cs

```diff
@@ -147,8 +147,11 @@ public DataType VisitBinaryExpression(BinaryExpression binExp)
             }
             else
                 throw NYI(binExp);
-            binExp.TypeVariable.DataType = dt;
-            binExp.TypeVariable.OriginalDataType = dt;
+            if (binExp.TypeVariable != null)
+            {
```

Comment:
I'd make `ExpressionTypeAscender` abstract, move it to `Reko.Core` and then have two subclasses, one for `TypeCollector` and one for `DataFlowAnalysis`